README.md
# Development and support ended
## Game v3.8 seems to work, no LazyQMG/fakeMail support.
**If there is any oppurtunity to update, it will be done, so thank you, and farewell!**<br>
For any questions, please contact me. [SlowsieNT#contact](https://github.com/SlowsieNT#contact)<br>
[Departure song](https://www.youtube.com/watch?v=MbiJ-VToYEg)<br>

# Genshin Impact Script(s)
## This is free and unencumbered software released into the public domain.
**Date(s)**: 08-jul-2023<br>
**Change(s)**: Unscheduled update<br>
**Status**: N/A<br>
## IMPORTANT
`LazyQMG` WILL cause issues (`Retrier3.ahk.inf`), please configure it, or disable it.<br>
Ladies and gentlemen, it's "My time has come" moment.

## Stat
<img src="https://api.visitorbadge.io/api/visitors?path=slowsient.genshin&countColor=%2337d67a&style=flat-square&labelStyle=upper">

## Contact?
See: [github.com/SlowsieNT#contact](https://github.com/SlowsieNT#contact)
(On github profile)


## Links / Download?
Accurate **GUIDE** for Retrier3 [Click here](https://github.com/SlowsieNT/Genshin-Impact/blob/main/Retrier3-Guided-Rerolling.md)<br>
**Please, check the guide because it will save you the trouble. Thank you.**<br>
Code > Download ZIP ([Try this link](https://github.com/SlowsieNT/Genshin-Impact/archive/refs/heads/main.zip))<br>
AutoHotkey v1.x ([Website link](https://www.autohotkey.com/))<br>
Violentmonkey Chrome ([Extension link](https://chrome.google.com/webstore/detail/violentmonkey/jinjaccalgkegednnccohejagnlnfdag)) or Vivaldi/Chromium...<br>
Violentmonkey Firefox ([Extension link](https://addons.mozilla.org/en-US/firefox/addon/violentmonkey/))<br>

## Available scripts:

🔗 __[Retrier3-Live](https://github.com/SlowsieNT/Genshin-Impact/tree/main/Retrier3-Live)__ is now available at GitHub!<br>
(Not a bot, it is Helper for questing)<br>
🔗 __[genshin-checkin-redeem.js](https://github.com/SlowsieNT/Genshin-Impact/blob/main/js/violentmonkey/genshin-checkin-redeem.js)__ is now available at GitHub!<br>
(Used for auto redeem check-in reward)<br>
🔗 __[quickfmg.js](https://github.com/SlowsieNT/Genshin-Impact/blob/main/js/violentmonkey/quickfmg.js)__ is now available at GitHub!<br>
(Used for auto generate random mail & copy it, and then copy verification code for mihoyo account)

## Retrier3
For Guide just [Click here](https://github.com/SlowsieNT/Genshin-Impact/blob/main/Retrier3-Guided-Rerolling.md)<br>
*(Unlike videos, guide is up-to-date)*

### FAQ
Q: What's center of the mouse?<br>
Q: What's center of the cursor?<br>
A: Look at [this picture (*LINK*)](https://github.com/SlowsieNT/genshin/blob/main/Visuals/center-of-cursor.png)<br><br>
Q: Will the script(s) be updated in the future?<br>
A: Yes - to keep it alive, but updates may be delayed in the future.<br><br>
Q: When will cease support take place?<br>
A: Anytime now.<br><br>
Q: Can it be ported over mobile?<br>
A: No, as of now there is no appropriate tool.<br><br>
Q: Can it be made to do entire reroll?<br>
A: It is very difficult to achieve such feat without dll injection, which is risky.<br><br>
Q: Is there any risk of using this script(s)?<br>
A: N/A. But yes if you get successfully blackmailed.<br><br>
Q: Why is tiktοk a sруwаrе?<br>
A: Brainwashing. Now back to the original topic.<br><br>
Q: Is there possibility of further improving the script?<br>
A: For now, there is none.<br>

## genshin-checkin-redeem.js
(requires Tampermonkey/Greasemonkey Opera/Chrome/Firefox extension)<br>
#### NOTE: Only the account that is currently logged in.<br>
Every 3h script will try to check-in<br>
If script asks to allow requests, click on "Always allow domain" or similar button.<br>
__Script will trigger when you are casually watching youtube videos, or just browsing__<br>
This script will send 2 requests, one will attempt to get checkin game reward, and the other request is another checkin<br>
I have tested it on other accounts and it should be working for you as well<br>

## quickfmg.js
(requires Tampermonkey/Greasemonkey Opera/Chrome/Firefox extension)<br>
Once you add this script and go to [emailfake.com](http://emailfake.com/), you would click on Logo that will have green background.<br>
Once you click on the Logo, it will copy the new generated mail automatically<br>
IMPORTANT: It will copy verification code for genshin impact account, no worries

## License
All of files are **Unlicense**d. No rights are reserved.<br>


CHANGELOG.md
## Repository
- Renamed from `genshin` to `Genshin-Impact`
- Still accessible at `SlowsieNT/Genshin`
## [js] checkin-redeem (Latest: 03-jun-2022)
#### (03-jun-2022)
- Add `@connect` to avoid "allow" popup.
#### (16-may-2022)
- No more irrelevant errors in dev console, yay!
## [js] quickfmg (Latest: 01-mar-2023)
Supported domains:
- [highly recommended] http://emailfake.com/
- [recommended] http://generator.email
- [not recommended] http://fakemailgenerator.com/
#### (01-mar-2023)
- Maintaining is now easier, again
- Debug function `DbgBulkValidate` now accepts ini string containing `MailTypeN` where `N` is index
- Remove expired mails
#### (18-jan, 02-jan-2023)
- (18-jan) Remove expired mails (remaining uptime less than 420days)
- (18-jan) Polish, allow easier debugging
- Add Stopwatch (starts from 6sec due taking 5-6sec for mail to arrive).
- Remove expired mails (remaining uptime less than 420days)
#### (08-dec-2022)
- Remove expired mails (remaining uptime less than 420days)
- Enables dbg functions if `location.hash` equals `#dbg`
#### (05-sep-2022)
- Remove expired mails (remaining uptime less than 420days)
- Add new mails
#### (19-aug, 17-aug-2022)
- (19-aug) Forgor to remove irrelevant code and do polish
- Add verify code copy for Tower of Fantasy!
- Remove expired mails (remaining uptime less than 366days)
#### (03-jun-2022)
- Update check #1, title of fake mail website will show "UPDATE" if script is updated (on github)
  - Message lasts for 5 seconds, title will reset.
- Update check #2, it may auto update script too, not tested
#### (01-jun-2022)
- Update invert filter of `html` (from 90% to 91%)
- Remove outdated mails (uptime less than ~200d)
- Add mails (uptime more than ~200d)
#### (16-may, 31-mar-2022)
- (16-may) Add invert filter to `html` (~90% invert)
- Add a lot of mails
- Add "load mail" feature
## Retrier3 (Latest: 08-jul-2023)
#### [Game: **v3.8**] (08-jul-2023)
- Readjust pixel hooks
- `LazyQMG` is disabled by default, also can use custom mail on register, at least `WM1440` is seen to work.
#### [Game: **v3.6**] (13-apr-2023)
- Readjust pixel hooks for `FS1920`, for now.
- Should fix `WM1440` at a later date, or at v3.7, no promises.
#### [Game: **v3.5**] (01-mar-2023)
- Readjust pixel hooks.
- Maintaining is now easier.
  - `LazyQMG` mails are now moved to `.ini` file.
- Remove expired mails from `LazyQMG`
#### [Game: **v3.4**] (18-jan-2023)
- Readjust pixel hooks.
- Again, remove expired mails from `LazyQMG` (remaining uptime less than 420days)
#### [Game: **v3.3**] (02-jan-2023)
- Readjust pixel hooks for Register form.
- Add update checker for Retrier3.ahk
- Again, remove expired mails from `LazyQMG` (remaining uptime less than 420days)
#### [Game: **v3.3**] (08-dec-2022)
- X-mas update (unscheduled).
- Readjusted pixel hooks
- Remove expired mails from `LazyQMG` (remaining uptime less than 420days)
#### [Game: **v3.2**]
- No updates were issued.
#### [Game: **v3.1**] (sep-2022)
- "They still don't care." (anniversary)
#### [Game: **v3.0**] (sep-2022)
- End of active support.
#### [Game: **v3.0**] (05-sep-2022)
- Yeet `mod.cutscene.ahk` as it is useless!
- Readjusted pixel hooks
- Added feature `Sounds` for `.inf`!
- Added `icecubes.mp3` for `Sounds` feature for people that reroll in pitch black room, so they can close eyes and wait for sound notification.
  - Finally we can reroll during nighttime with less eyedamage!
- Remove expired mails from `LazyQMG` (remaining uptime less than 420days)
- Add new mails to `LazyQMG`
#### [Game: **v3.0**] (24-aug-2022)
- Readjusted pixel hook(s)
  - `UIIGOpenedJournal`
#### [Game: **v2.8**] (17-aug-2022)
- Remove expired mails (remaining uptime less than 366days)
#### [Game: **v2.8**] (13-jul-2022)
- Readjusted pixel hook(s)
  - Too many nerfs to write
#### [Game: **v2.7**] (01-jul-2022)
- Edit github repo logo
- Remove expired mails from `LazyQMG`
  - Mails with less than ~221 days are removed
- Fix: Script preferred default value over ini defined value
  - The `if (Trim(str))` if string were `0` ahk would assume as false value, thus caused the issue
    - Added the right logic
#### [Game: **v2.7**] (08-jun-2022)
- Reduce clicks necessary to redeem code `genshingift`
#### [Game: **v2.7**] (06-jun-2022)
- Add feature to generate mail upon pressing F2 at registration form
  - Can be disabled in `.inf` by setting section `LazyQMG` field `Allow` to `0`
- When prompted with "Press Space" it will now always plunge
  - Can be disabled in `.inf` by setting section `HiddenSettings` field `PressSpaceOnce` to `1`
#### [Game: **v2.7**] (01-jun-2022)
- Add feature to auto "Press 2" when prompted (long lost Retrier3 code; back in the days of Retrier2)
- Add feature to auto "Press W"
- Add feature to instantly plunge on "Press Space"
- Add feature after arriving to *castle* to hold W
- Add feature after auto pressing W, continues to hold W
#### [Game: **v2.7**] (01-Jun, 31-may-2022)
- Readjusted pixel hook(s)
  - Hook(s): `UIIGAutoplay`, `UIIGUndiscoveredWaypoint`, `UIIGOpenedJournal`
  - Hook(s): `UIIGUndiscoveredWaypoint` (01-Jun)
- Readjusted: Pos hook diag opt
#### [Game: **v2.6**] (16-may-2022)
- Readjusted pixel hook(s)
  - Hook `UIIGUndiscoveredWaypoint`
#### [Game: **v2.6**] (30-mar-2022)
- Readjusted pixel hook(s)
  - Register, InGame UI
- Removed ONLY FEW mail domains that fail on emailfake
#### [Game: **v2.5**] (16-feb-2022)
- Readjusted pixel hook(s) for `WM1440`
- Readjusted pixel hook(s) for `FS1920`
- Readjusted when mail is logged.
- Readjusted when "Lv5" mail is logged.
- Added message if "Windowed 1440x900" is NOT supported.
#### [Game: **v2.4**] (12-feb-2022)
- Features to `Retrier.ahk.inf`
  - Readjusted logic for `LoggingMail`
    - Mails will be logged to file after blackscreen if `Allow` is set to `1`
  - Added `LoggingMail` section, by field `AfterFOTeleport` to allow logging
    - Triggers if script was running during "fend off"

js/tampermonkey.genshin-redeem.js
// ==UserScript==
// @name         Auto Redeem01
// @namespace    http://tampermonkey.net/
// @version      0.1
// @description  try to take over the world!
// @author       SlowsieNT
// @match        *://*/*
// @grant GM_getValue
// @grant GM_setValue
// @grant unsafeWindow
// ==/UserScript==

(function () {
	'use strict';
	var vScrPrefix = "dasadsnuf_2021_date", vInterval = 3 * 3600e3;
	var vLastDate = GM_getValue(vScrPrefix);
	var vDiff = Date.now() - vLastDate;
	var vACT_ID = "e202102251931481";
	// try fix if failure occurs
	if (!vDiff && 0 == vLastDate >> 0)
		GM_setValue(vScrPrefix, Date.now()), send_signin();
	// if diff is greater than 8h, then try signin
	if (vDiff > vInterval)
		GM_setValue(vScrPrefix, Date.now()), send_signin()
	function send_signin() {
		// THIS PART of the code claims daily reward for GAME
		fetch("https://hk4e-api-os.mihoyo.com/event/sol/sign?lang=en-us", {
			"headers": {
				"accept": "application/json, text/plain, */*",
				"accept-language": "en-US,en;q=0.9",
				"content-type": "application/json;charset=UTF-8",
				"sec-fetch-dest": "empty",
				"sec-fetch-mode": "cors",
				"sec-fetch-site": "same-site"
			},
			"referrer": "https://webstatic-sea.mihoyo.com/ys/event/signin-sea/index.html?act_id="+vACT_ID+"&lang=en-us",
			"referrerPolicy": "no-referrer-when-downgrade",
			"body": "{\"act_id\":\""+vACT_ID+"\"}",
			"method": "POST",
			"mode": "cors",
			"credentials": "include"
		});
		// THIS PART of the code claims another check-in
		fetch("https://api-os-takumi.mihoyo.com/community/apihub/api/signIn", {
			"headers": {
				"accept": "application/json, text/plain, */*",
				"accept-language": "en-US,en;q=0.9",
				"content-type": "application/json;charset=UTF-8",
				"sec-fetch-dest": "empty",
				"sec-fetch-mode": "cors",
				"sec-fetch-site": "cross-site",
				"x-rpc-language": "en-us"
			},
			"referrer": "https://www.hoyolab.com/",
			"referrerPolicy": "strict-origin-when-cross-origin",
			"body": "{\"gids\":\"2\"}",
			"method": "POST",
			"mode": "cors",
			"credentials": "include"
		});
	}
})();


js/tampermonkey/genshin-checkin-redeem.js
// ==UserScript==
// @name         Auto Redeem01
// @namespace    http://tampermonkey.net/
// @version      0.1
// @description  try to take over the world!
// @author       SlowsieNT
// @match        *://*/*
// @connect hk4e-api-os.mihoyo.com
// @connect api-os-takumi.mihoyo.com
// @grant GM_xmlhttpRequest
// @grant GM_getValue
// @grant GM_setValue
// @grant unsafeWindow
// ==/UserScript==

(function () {
	'use strict';
	var vScrPrefix = "scr_J202205_date", vInterval = 3 * 3600e3;
	var vLastDate = GM_getValue(vScrPrefix);
	var vDiff = Date.now() - vLastDate;
	var vACT_ID = "e202102251931481";
	// try fix if failure occurs
	if (!vDiff && 0 == vLastDate >> 0)
		GM_setValue(vScrPrefix, Date.now()), Checkin();
	// if diff is greater than (x)h, then try signin
	if (vDiff > vInterval)
		GM_setValue(vScrPrefix, Date.now()), Checkin();
	function Checkin() {
		// THIS PART of the code claims daily reward for GAME
		GM_xmlhttpRequest({
			method: "POST", url: "https://hk4e-api-os.mihoyo.com/event/sol/sign?lang=en-us",
			data: "{\"act_id\":\""+vACT_ID+"\"}",
			headers: {
				"accept": "application/json, text/plain, */*",
				"accept-language": "en-US,en;q=0.9",
				"content-type": "application/json;charset=UTF-8",
				"referrer": "https://webstatic-sea.mihoyo.com/ys/event/signin-sea/index.html?act_id="+vACT_ID+"&lang=en-us",
				"referrerPolicy": "no-referrer-when-downgrade",
			}
		});
		// Community xp check-in
		GM_xmlhttpRequest({
			method: "POST", url: "https://api-os-takumi.mihoyo.com/community/apihub/api/signIn",
			data: "{\"gids\":\"2\"}",
			headers: {
				"accept": "application/json, text/plain, */*",
				"accept-language": "en-US,en;q=0.9",
				"content-type": "application/json;charset=UTF-8",	
				"x-rpc-language": "en-us"
			}
		});
	}
})();


js/violentmonkey/quickfmg.js
// ==UserScript==
// @name         Quick FakeMailGen
// @namespace    http://tampermonkey.net/
// @version      0.1.3
// @description  try to take over the world!
// @author       You
// @match        *://*.fakemailgenerator.com/*
// @match        *://*.generator.email/*
// @match        *://*.emailfake.com/*
// @icon         https://www.google.com/s2/favicons?domain=fakemailgenerator.com
// @grant        unsafeWindow
// @grant        GM_setClipboard
// @grant        GM_addStyle
// @grant        GM_info
// @grant        GM_xmlhttpRequest
// @run-at       document-start
// @connect      github.com
// @connect      raw.githubusercontent.com
// @require      https://ajax.googleapis.com/ajax/libs/jquery/1.7.2/jquery.min.js
// ==/UserScript==

(function() {
	'use strict';
	// remove the flashbang:
	GM_addStyle('html{filter:invert(.91);}img,.navbar-brand,.nav-link{filter:invert(.91);}');
	// do not edit below unless dev:
	var LAST_UPDATE = 1677678402, LiveLU, CanCheck = 0, EPType = -1,
		ScriptName = GM_info.script.name, LHost = location.hostname;
	function UpdateAvailable() {
		return (0||LiveLU) > LAST_UPDATE;
	}
	function RandomUpdateTime() {
		return Date.now() + Math.random()*300e3 + 60e3;
	}
	function GetDHMS(aTime) {
		var d = 0|aTime/86400,
			h = 0|aTime/3600%24,
			m = 0|aTime/60%60,
			s = 0|aTime%60,
			ms = 0|aTime%1*1000;
		if (d < 10) d = "0" + d;
		if (h < 10) h = "0" + h;
		if (m < 10) m = "0" + m;
		if (s < 10) s = "0" + s;
		return d + ":" + h + ":" + m + ":" + s + "." + ms;
	}
	function RRStopwatch() {
		// -6e3 because it usually takes ~6sec for verify code to arrive
		var epoch = Date.now() - 6e3;
		setInterval(function () {
			document.title = GetDHMS(  ((Date.now()-epoch)/1e3).toFixed(3)  );
		}, 1000/60);
	}
	function UpdateCheck(aFunc) {
		var vURL = "https://raw.githubusercontent.com/SlowsieNT/Genshin-Impact/main/Retrier3-Live/Etc.Updates.ini";
		GM_xmlhttpRequest({
			url: vURL,
			onload: function (aX) {
				if (200 == aX.status) {
					var vVer = aX.response.split(/date=(.+)[\r|\n]+/g).filter(Boolean);
					if (1 in vVer)
						vVer = +vVer[1];
					else vVer = 0;
					LiveLU = vVer;
					aFunc();
				}
			}
		});
	}
	if (-1 < LHost.indexOf("fakemailgenerator.com")) CanCheck=1,EPType=0;
	if (-1 < LHost.indexOf("generator.email")) CanCheck=1,EPType=1;
	if (-1 < LHost.indexOf("emailfake.com")) CanCheck=1,EPType=2;
	CanCheck && UpdateCheck(function () {
		$(function() {
			var vOT = document.title;
			if (UpdateAvailable()) {
				document.title = "UPDATE UserScript: " + ScriptName;
				setTimeout(function () { document.title = "UPDATE: " + ScriptName; }, 1500);
				setTimeout(function () {
					document.title = vOT;
				}, 5e3);
			}
		});
	});
	$(function(){
		if (0 === EPType) FakeMailGeneratorCom();
		else if (1 === EPType) GeneratorEmail();
		else if (2 === EPType) EmailFake();
		GM_setClipboard(($("#email_ch_text").text()||$("#cxtEmail").text()||"").trim());
		RunVCHandlers();
	});
	function RunVCHandlers() {
		var vTmrChk2 = setInterval(function () {
			if (-1 === document.title.toLowerCase().indexOf("tower of fantasy"))
				return;
			// brute force the code location
			var codes = Array.apply(0, $("#email-table *")).map(function(y, x){
				// assume x positive, and length less than 7
				if (+x && 7 > (x = y.innerText.trim()).length)
					return +x;
			}).filter(function(x){return x;});
			if (0 in codes)
				GM_setClipboard(codes[0]), clearInterval(vTmrChk2);
		}, 2e2);
		// copy mail handler:
		var vTmrChk1 = setInterval(function () {
			var vParagraph = $("[class*=subj_div]:contains(ifica)");
			if (vParagraph.length)
				GM_setClipboard(vParagraph.text().split(" ")[0]), clearInterval(vTmrChk1),RRStopwatch();
		}, 2e2);
	}
	var Generator = new function () {
		var _ = this;
		function RandomStr(aLength) {
			// couldn't be bothered to use one Random() call, yes
			for (var i = 0, r = ""; i < aLength; i++) {
				// we use bitwise flex for flooring number in range: 2147483647, -2147483648
				var t = [48,97][Math.random()*2>>0];
				// legend says ternary operator is slower than "if", couldn't be bothered again
				r += t>48?String.fromCharCode(t+Math.random()*6>>0):Math.random()*10>>0;
			}
			// people define variables outside "for", oh no
			return r;
		}
		_.GenerateMail0 = function () {
			function GenerateMD(x) {
				return (_.Mails0)[_.Mails0.length*Math.random()>>0];
			}
			var vDomain = GenerateMD(),
					vLengthRng = 8+Math.random()*16>>0,
					vFancyLetter = String.fromCharCode(97+6+Math.random()*16>>0),
					vUsername = RandomStr(vLengthRng)+vFancyLetter+Date.now();
			return {
				Name: vUsername, Host: vDomain,
				URL: 'http://www.fakemailgenerator.com/inbox/' + vDomain.slice(1) + '/' + vUsername + '/',
				toString: function () {
					return vUsername+"@"+vDomain;
				}
			}
		};
		_.GenerateMail1 = function () {
			function GenerateMD() {
				// removed expired mails
				return _.Mails1[_.Mails1.length*Math.random()>>0];
			}
			var vDomain = GenerateMD(),
					vLengthRng = 4+Math.random()*4>>0,
					vFancyLetter = String.fromCharCode(97+6+Math.random()*16>>0),
					vUsername = RandomStr(vLengthRng)+vFancyLetter+Date.now();
			return {
				Name: vUsername, Host: vDomain,
				toString: function () {
					return vUsername+"@"+vDomain;
				}
			}
		};
		_.NavigateMail1 = function (aEmail) {
			if ("string" === typeof aEmail && aEmail.trim())
				location.href = 'https://generator.email/' + aEmail;
		};
		_.GenerateMail2 = function () {
			function GenerateMD(x) {
				// removed expired mails
				return (
					_.Mails2)[_.Mails2.length*Math.random()>>0];
			}
			var vDomain = GenerateMD(),
					vLengthRng = 4+Math.random()*4>>0,
					vFancyLetter = String.fromCharCode(97+6+Math.random()*16>>0),
					vUsername = RandomStr(vLengthRng)+vFancyLetter+Date.now();
			return {
				Name: vUsername, Host: vDomain,
				URL: "https://emailfake.com/"+vDomain+"/"+vUsername,
				toString: function () {
					return vUsername+"@"+vDomain;
				}
			}
		};
		_.NavigateMail2 = function (aEmail) {
			if ("string" === typeof aEmail && aEmail.trim()) {
				var vParts = aEmail.split("@");
				location.href = "https://emailfake.com/" + vParts[1] + "/" + vParts[0];
			}
		};
		_.Mails0 = (
			["armyspy.com", "cuvox.de",
				"dayrep.com", "einrot.com", "fleckens.hu",
				"gustr.com", "jourrapide.com", "rhyta.com",
				"superrito.com", "teleworm.us"]
		);
		_.Mails1 = (
			["gmailvn.net","enhanceronly.com","alvinneo.com","colevillecapital.com","plexvenet.com","pickuplanet.com","ebarg.net","btcmod.com","pianoxltd.com","filevino.com","posthectomie.info","24hinbox.com","bizimalem-support.de","asifboot.com","ma2limited.com","shurkou.com","bizisstance.com","boranora.com","email-temp.com","packiu.com","hotmail.red","nudinar.net","googl.win","crossfitcoastal.com","greendike.com","tubidu.com","emvil.com","pow-pows.com","onlyu.link","dmxs8.com","chantellegribbon.com","stinkypoopoo.com","rackabzar.com","guitarsxltd.com","netveplay.com","nx1.us","onlinecmail.com","saxophonexltd.com","goliszek.net","arss.me","yt-google.com","degar.xyz","816qs.com","cggup.com","speeddataanalytics.com","skillion.org","falixiao.com","getcashstash.com"]
		);
		_.Mails2 = (
			["chantellegribbon.com","24hinbox.com","alvinneo.com","freeallapp.com","packiu.com","emvil.com","goldinbox.net","colevillecapital.com","kenvanharen.com","luddo.me","bizisstance.com","boranora.com","hotmail.red","rackabzar.com","dmxs8.com","pickuplanet.com","24mail.top","nudinar.net","filevino.com","greendike.com","getcashstash.com","gmailvn.net","asifboot.com","tubidu.com","plexvenet.com","stinkypoopoo.com","googl.win","happiseektest.com","enhanceronly.com","skillion.org","asistx.net","posthectomie.info","pianoxltd.com","guitarsxltd.com","ebarg.net","email-temp.com","onlinecmail.com","btcmod.com","falixiao.com","yt-google.com","fuadd.me","muvilo.net","saxlift.us","cggup.com","crossfitcoastal.com","livegolftv.com","cashbackr.com","stamberg.nl","coffeepancakewafflebacon.com"]
		);
	};
	// for: generator.email
	function GeneratorEmail() {
		// once user click on logo, it will redirect to random email link
		$(".navbar-brand").click(function () {
			var vMail = Generator.GenerateMail1();
			GM_setClipboard("" + vMail);
			// navigate to new mail
			location.href = 'https://generator.email/' + vMail;
		}).css("background-color", "green").attr("href", "#");
		var vNL = $(".nav-link:nth(0)");
		$("span:last", vNL).text("Load mail...");
		vNL.click(function () {
			Generator.NavigateMail1(prompt("Enter mail to load:"));
		}).css("background-color", "green").attr("href", "#");
	}
	// for: fakemailgenerator.com
	function FakeMailGeneratorCom() {
		// once user click on logo, it will redirect to random email link
		$(".navbar-brand").click(function () {
			var vMail = Generator.GenerateMail0();
			GM_setClipboard("" + vMail);
			// navigate to new mail
			location.href = vMail.URL;
		}).css("background-color", "green").attr("href", "#");
	}
	function EmailFake(){
		// once user click on logo, it will redirect to random email link
		$(".navbar-brand").click(function () {
			var vMail = Generator.GenerateMail2();
			GM_setClipboard("" + vMail);
			location.href = vMail.URL;
		}).css("background-color", "green").attr("href", "#");
		// prompt to load mail by raw email
		var vNL = $(".nav-link:nth(0)");
		$("span:last", vNL).text("Load mail...");
		vNL.click(function () {
			Generator.NavigateMail2(prompt("Enter mail to load:"));
		}).css("background-color", "green").attr("href", "#");
	}
	function DbgFetchMails(aFunc, aErrFunc) {
		$.get("/", function (aResp) {
			var d = new DOMParser().parseFromString(aResp, "text/html");
			var arr = Array.apply(0, $("[onclick*=change_dropdown_list]", d)).map(function (a) {
				return a.innerText.trim();
			});
			aFunc(arr);
		}).error(aErrFunc);
	}
	// imagine i added a function useful only to devs
	function DbgValidate(aFunc, aErrFunc, aUsr, aDmn, aDays) {
		$.post("/check_adres_validation3.php", "usr="+encodeURIComponent(aUsr)+"&dmn="+encodeURIComponent(aDmn)).done(function (aResp) {
			var jo = JSON.parse(aResp);
			aFunc(jo.uptime > (aDays||420), jo);
		}).error(function () {
			aErrFunc.apply(this, arguments);
		});
	}
	// IMAGINE... REAL SHET
	function DbgBulkValidate(aFunc, aItems, aIniIndex, aDays) {
		// use aIni=1 and aIndex bigger than 0 then aItems is ini string of
		// MailType0=... MailType1=... and MailType2=...
		var oItems = aItems,
			hasII = -1 < aIniIndex,
			isStr = "string" === typeof aItems;
		if (isStr && hasII)
			aItems = aItems.split(",");
		else if (isStr && hasII)
			aItems = aItems.split("MailType"+aIniIndex+"=")[1].split(/\n|\r/)[0].trim().split(",");
		var ret = [], cd = 0, elen = aItems.length;
		// 'elen, my beloved where r u?
		// elen's bitch, ahem, expected length
		function doneFn(ret) {
			if (hasII) {
				var t, lines = oItems.split(/\r\n|\r|\n/);
				console.log("[dbg] aIniIndex =", aIniIndex);
				for (var i in lines)
					if (lines[i].trim().startsWith(t = "MailType" + aIniIndex))
						lines[i] = t + "=" + ret;
				console.log("[dbg] new lines =", lines);
				return aFunc(lines.join("\n"));
			}
			return aFunc(ret);
		}
		(function Next(aIndex) {
			var item = aItems[aIndex];
			if (item) DbgValidate(function (isok) {
				if (isok && -1 === ret.indexOf(item))
					ret.push(item);
				if (++cd >= elen)
					doneFn(ret);
			}, function () {
				if (++cd >= elen)
					doneFn(ret);
			}, "a", item, aDays);
			// stack em async XHRs
			if (item)
				setTimeout(Next, 255, ++aIndex);
		})(0);
	}
	if ("#dbg" == location.hash) {
		// DbgBulkValidate(x=>console.log(x), Generator.Mails1)
		unsafeWindow.Generator=Generator;
		unsafeWindow.DbgFetchMails=DbgFetchMails;
    	unsafeWindow.DbgBulkValidate=DbgBulkValidate;
		console.log(RandomUpdateTime());
	}
})();
